File: cpp_10-99/17.cc
// C++
// leetcode 17
// url: https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/
// 思路:
// 将ret写在类里，然后写一个dfs的函数，用dfs把答案装进ret里

class Solution {
public:
    vector<string> phone_keyboard{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    vector<string> ret;

    void dfs(string pre, int *num, int done,int size)
    {
        if(size ==done)
        {
            ret.push_back( pre);
            return;
        }
        for(char c: phone_keyboard[ num[done]])
        {
            string temp = pre + c;
            dfs(temp, num, done+1, size);
        }
    }

    vector<string> letterCombinations(string digits)
    {
        int size = digits.size();
        if(size ==0)
            return ret;
        int num[size];
        for(int i =0; i<size; i++)
        {
            num[i] = ((int)digits[i] -50);
        }
        string temp[size];
        dfs("", num, 0, size);
        return ret;
    }
};

File: Go/Go_10-99/17.go
// Go
// leetcode 17
// https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number/

var getBytes []string = []string{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}

var res []string
func dfs(s string, cur int, arr []rune) {
	if cur == len(arr) {
		res = append(res, string(arr))
		return
	}

	for _, b := range getBytes[int(s[cur] - '2')] {
		arr[cur] = b
		dfs(s, cur + 1, arr)
	}
}

func letterCombinations(digits string) []string {
	res = []string{}
	n := len(digits)
    if n == 0 {
        return res
    }
	tmp := make([]rune, n)
	dfs(digits, 0, tmp)
	return res
}